### PR TITLE
[11.x] fix: Model newCollection generics; feat: add HasCollection trait

### PR DESCRIPTION
**Pull request by calebdw**
Hello!

This PR fixes some issues with the Model's newCollection generic (https://github.com/phpstan/phpstan/issues/11354) ---it appears that due to covariant/invariant constraints the method must return only static models.

I also added a `HasCollection` trait similar to the `Has{Builder,Factory}` traits that help slim down the models:
```php
// before
class Post extends Model
{
    /** @return PostCollection<array-key, static> */
    public function newCollection(array $models = []): PostCollection
    {
        return new PostCollection($models);
    }
}

// after
class Post extends Model
{
    /** @use HasCollection<PostCollection<array-key, static>> */
    use HasCollection;

    protected static string $collection = PostCollection::class;
}
```

Thanks!

### CODE REVIEW

**Comment by henzeb** [comment]
Looking at your code, you add a trait, but then you aren't doing anything with the trait, it has zero effect, the newCollection is still in Model.

**Comment by calebdw** [comment]
The trait is for use in user land when a custom collection class is used as it better helps with generics/autocomplete/ide support, however, I suppose we could add the trait to the base model like this:

https://phpstan.org/r/0c35c490-55a1-48fa-b407-d58ae52cb499

**Comment by henzeb** [comment]
> The trait is for use in user land when a custom collection class is used as it better helps with generics/autocomplete/ide support

if it is just for that, it doesn't slim down anything. I suggest either remove the trait or move generic logic into that trait. leaving Model as was.

**Comment by henzeb** [comment]
The trait is still not doing anything except for extending the original newCollection method. The trait should not have to be part of Model by default. by default Model uses the default Collection class. But only when adding that trait, one should be able to set a specific Collection classname. No code changes in Model, just a new trait.

**Comment by calebdw** [comment]
Thanks @taylorotwell!

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -218,6 +220,13 @@ abstract class Model implements Arrayable, ArrayAccess, CanBeEscapedWhenCastToSt
      */
     protected static string $builder = Builder::class;
 
+    /**
+     * The Eloquent collection class to use for the model.
+     *
+     * @var class-string<\Illuminate\Database\Eloquent\Collection<*, *>>
+     */
+    protected static string $collection = Collection::class;
```
Comment (GromNaN) [inline src/Illuminate/Database/Eloquent/Model.php:228]: Unfortunately this property name is already used by a popular third party package. I opened a PR to propose a new name: https://github.com/laravel/framework/pull/52186

Comment (calebdw) [reply src/Illuminate/Database/Eloquent/Model.php:228]: I wasn't aware, thanks!

Comment (GromNaN) [reply src/Illuminate/Database/Eloquent/Model.php:228]: No problem, I hope the patch will be accepted before this feature is released.

Comment (calebdw) [reply src/Illuminate/Database/Eloquent/Model.php:228]: Just out of curiosity, why don't they just use the existing `$table` property for that purpose?

Comment (GromNaN) [reply src/Illuminate/Database/Eloquent/Model.php:228]: In the MongoDB terminology, a collection is the equivalent of an SQL table (see [this doc](https://www.mongodb.com/docs/manual/reference/sql-comparison/).

This property was introduced [a long time](https://github.com/mongodb/laravel-mongodb/commit/7effee857b64b2fb1d499e2452446661d568a7bf) before I start working on the project. I think they wanted to use the terms in the code.

**Pull request merged**